### PR TITLE
CR for job #21142 Create a rhythm practice toy

### PR DESCRIPTION
**Pull request by TopRamenGod**
Code Review for Job #20924 - Workitem available at https://worklist.net/21142

Notes: 
- Despite numerous avenues of solution, timing is still inconsistent at higher BPMs.
- In the worklist description, there is a request, "While playing, if the hand controller again intersects the sphere, play a slightly different percussion sound." This was attempted, but due to tight timing, the sound never fires in a window narrow enough to actually 'play' to the user, as so this functionality was removed.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 46bd78070a76094b0b2d7d689fe4b07692d77cab available at: https://gist.github.com/0ff34e6f9edfb9fcb81800480f0b673e<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9732-46bd780.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9732-46bd780.dmg

**Comment by cozza13** [comment]
ok to test

**Comment by hifi-gustavo** [comment]
Build console output for commit: 46bd78070a76094b0b2d7d689fe4b07692d77cab available at: https://gist.github.com/0ff34e6f9edfb9fcb81800480f0b673e<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9732-46bd780.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9732-46bd780.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: ee3a30eee93508730b81dd56eedacdaf6cac151e available at: https://gist.github.com/0ff34e6f9edfb9fcb81800480f0b673e<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9732-ee3a30e.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9732-ee3a30e.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 32ea431fdf4921fca8c57af7006b898addc15098 available at: https://gist.github.com/0ff34e6f9edfb9fcb81800480f0b673e<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9732-32ea431.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9732-32ea431.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 75897af655600ec6a838a84084650989558e72b4 available at: https://gist.github.com/0ff34e6f9edfb9fcb81800480f0b673e<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9732-75897af.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9732-75897af.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: d3ea79c810b1b53a41231819131818ff51003f87 available at: https://gist.github.com/0ff34e6f9edfb9fcb81800480f0b673e<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9732-d3ea79c.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9732-d3ea79c.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2d5b9bffa87e275e7a1f523f14511b2576136cca available at: https://gist.github.com/0ff34e6f9edfb9fcb81800480f0b673e<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9732-2d5b9bf.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9732-2d5b9bf.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 1235ae9e178eb0a7aa6869b999d270a48f5e6532 available at: https://gist.github.com/0ff34e6f9edfb9fcb81800480f0b673e<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9732-1235ae9.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9732-1235ae9.dmg

**Comment by PhilipRosedale** [comment]
On testing latest... on starting, I get several beats and then a siren repeating forever with log output:


[2/2 22:30:37] [Entities 1] get blank padded line!!!!!!
[2/2 22:30:37] [Entities 1] REQUESTING PADDED LINE 0
[2/2 22:30:37] [Entities 1] get blank padded line!!!!!!
[2/2 22:30:37] [Entities 1] REQUESTING PADDED LINE 1
[2/2 22:30:37] [Entities 1] get blank padded line!!!!!!
[2/2 22:30:37] [Entities 1] REQUESTING PADDED LINE 2
[2/2 22:30:37] [Entities 1] get blank padded line!!!!!!
[2/2 22:30:37] [Entities 1] REQUESTING PADDED LINE 3
[2/2 22:30:37] [Entities 1] get blank padded line!!!!!!
[2/2 22:30:37] [Entities 1] REQUESTING PADDED LINE 4
[2/2 22:30:37] [Entities 1] get blank padded line!!!!!!
[2/2 22:30:37] [Entities 1] REQUESTING PADDED LINE 5
[2/2 22:30:37] [Entities 1] get blank padded line!!!!!!
[2/2 22:30:37] [Entities 1] REQUESTING PADDED LINE 6
[2/2 22:30:37] [Entities 1] get blank padded line!!!!!!
[2/2 22:30:37] [Entities 1] REQUESTING PADDED LINE 7
[2/2 22:30:37] [Entities 1] get blank padded line!!!!!!
[2/2 22:30:37] [Entities 1] getNewScoreboardDisplay iteration: 0
[2/2 22:30:37] [Entities 1] newScoreboardLinePosition
[2/2 22:30:37] [Entities 1] undefined
[2/2 22:30:37] [Entities 1] ERROR - [UncaughtException] TypeError: Result of expression 'newScoreboardLinePosition' [undefined] is not an object. in Entities 1:575
[Backtrace]
    <anonymous>() at http://theblacksun.s3.amazonaws.com/props/beatMatcher/beatMatcher.js:575
    <anonymous>(type = 'normal') at http://theblacksun.s3.amazonaws.com/props/beatMatcher/beatMatcher.js:391
    <anonymous>() at http://theblacksun.s3.amazonaws.com/props/beatMatcher/beatMatcher.js:1112
    <anonymous>() at http://theblacksun.s3.amazonaws.com/props/beatMatcher/beatMatcher.js:989
    <global>() at -1
[2/2 22:30:38] [Entities 1] get blank padded line!!!!!!
[2/2 22:30:38] [Entities 1] REQUESTING PADDED LINE 0
[2/2 22:30:38] [Entities 1] get blank padded line!!!!!!
[2/2 22:30:38] [Entities 1] REQUESTING PADDED LINE 1
[2/2 22:30:38] [Entities 1] get blank padded line!!!!!!
[2/2 22:30:38] [Entities 1] REQUESTING PADDED LINE 2
[2/2 22:30:38] [Entities 1] get blank padded line!!!!!!
[2/2 22:30:38] [Entities 1] REQUESTING PADDED LINE 3
[2/2 22:30:38] [Entities 1] get blank padded line!!!!!!
[2/2 22:30:38] [Entities 1] REQUESTING PADDED LINE 4
[2/2 22:30:38] [Entities 1] get blank padded line!!!!!!
[2/2 22:30:38] [Entities 1] REQUESTING PADDED LINE 5
[2/2 22:30:38] [Entities 1] get blank padded line!!!!!!
[2/2 22:30:38] [Entities 1] REQUESTING PADDED LINE 6
[2/2 22:30:38] [Entities 1] get blank padded line!!!!!!
[2/2 22:30:38] [Entities 1] REQUESTING PADDED LINE 7
[2/2 22:30:38] [Entities 1] get blank padded line!!!!!!
[2/2 22:30:38] [Entities 1] getNewScoreboardDisplay iteration: 0
[2/2 22:30:38] [Entities 1] newScoreboardLinePosition
[2/2 22:30:38] [Entities 1] undefined
[2/2 22:30:38] [Entities 1] ERROR - [UncaughtException] TypeError: Result of expression 'newScoreboardLinePosition' [undefined] is not an object. in Entities 1:575
[Backtrace]
    <anonymous>() at http://theblacksun.s3.amazonaws.com/props/beatMatcher/beatMatcher.js:575
    <anonymous>(type = 'normal') at http://theblacksun.s3.amazonaws.com/props/beatMatcher/beatMatcher.js:391
    <anonymous>() at http://theblacksun.s3.amazonaws.com/props/beatMatcher/beatMatcher.js:1112
    <anonymous>() at http://theblacksun.s3.amazonaws.com/props/beatMatcher/beatMatcher.js:989
    <global>() at -1
[2/2 22:30:39] [Entities 1] get blank padded line!!!!!!
[2/2 22:30:39] [Entities 1] REQUESTING PADDED LINE 0
[2/2 22:30:39] [Entities 1] get blank padded line!!!!!!
[2/2 22:30:39] [Entities 1] REQUESTING PADDED LINE 1
[2/2 22:30:39] [Entities 1] get blank padded line!!!!!!
[2/2 22:30:39] [Entities 1] REQUESTING PADDED LINE 2
[2/2 22:30:39] [Entities 1] get blank padded line!!!!!!
[2/2 22:30:39] [Entities 1] REQUESTING PADDED LINE 3
[2/2 22:30:39] [Entities 1] get blank padded line!!!!!!
[2/2 22:30:39] [Entities 1] REQUESTING PADDED LINE 4
[2/2 22:30:39] [Entities 1] get blank padded line!!!!!!
[2/2 22:30:39] [Entities 1] REQUESTING PADDED LINE 5
[2/2 22:30:39] [Entities 1] get blank padded line!!!!!!
[2/2 22:30:39] [Entities 1] REQUESTING PADDED LINE 6
[2/2 22:30:39] [Entities 1] get blank padded line!!!!!!
[2/2 22:30:39] [Entities 1] REQUESTING PADDED LINE 7
[2/2 22:30:39] [Entities 1] get blank padded line!!!!!!
[2/2 22:30:39] [Entities 1] getNewScoreboardDisplay iteration: 0
[2/2 22:30:39] [Entities 1] newScoreboardLinePosition
[2/2 22:30:39] [Entities 1] undefined
[2/2 22:30:39] [Entities 1] ERROR - [UncaughtException] TypeError: Result of expression 'newScoreboardLinePosition' [undefined] is not an object. in Entities 1:575
[Backtrace]
    <anonymous>() at http://theblacksun.s3.amazonaws.com/props/beatMatcher/beatMatcher.js:575
    <anonymous>(type = 'normal') at http://theblacksun.s3.amazonaws.com/props/beatMatcher/beatMatcher.js:391
    <anonymous>() at http://theblacksun.s3.amazonaws.com/props/beatMatcher/beatMatcher.js:1112
    <anonymous>() at http://theblacksun.s3.amazonaws.com/props/beatMatcher/beatMatcher.js:989
    <global>() at -1
[2/2 22:30:40] [Entities 1] get blank padded line!!!!!!
[2/2 22:30:40] [Entities 1] REQUESTING PADDED LINE 0
[2/2 22:30:40] [Entities 1] get blank padded line!!!!!!
[2/2 22:30:40] [Entities 1] REQUESTING PADDED LINE 1
[2/2 22:30:40] [Entities 1] get blank padded line!!!!!!
[2/2 22:30:40] [Entities 1] REQUESTING PADDED LINE 2
[2/2 22:30:40] [Entities 1] get blank padded line!!!!!!
[2/2 22:30:40] [Entities 1] REQUESTING PADDED LINE 3
[2/2 22:30:40] [Entities 1] get blank padded line!!!!!!
[2/2 22:30:40] [Entities 1] REQUESTING PADDED LINE 4
[2/2 22:30:40] [Entities 1] get blank padded line!!!!!!
[2/2 22:30:40] [Entities 1] REQUESTING PADDED LINE 5
[2/2 22:30:40] [Entities 1] get blank padded line!!!!!!
[2/2 22:30:40] [Entities 1] REQUESTING PADDED LINE 6
[2/2 22:30:40] [Entities 1] get blank padded line!!!!!!
[2/2 22:30:40] [Entities 1] REQUESTING PADDED LINE 7
[2/2 22:30:40] [Entities 1] get blank padded line!!!!!!
[2/2 22:30:40] [Entities 1] getNewScoreboardDisplay iteration: 0
[2/2 22:30:40] [Entities 1] newScoreboardLinePosition
[2/2 22:30:40] [Entities 1] undefined
[2/2 22:30:40] [Entities 1] ERROR - [UncaughtException] TypeError: Result of expression 'newScoreboardLinePosition' [undefined] is not an object. in Entities 1:575
[Backtrace]
    <anonymous>() at http://theblacksun.s3.amazonaws.com/props/beatMatcher/beatMatcher.js:575
    <anonymous>(type = 'normal') at http://theblacksun.s3.amazonaws.com/props/beatMatcher/beatMatcher.js:391
    <anonymous>() at http://theblacksun.s3.amazonaws.com/props/beatMatcher/beatMatcher.js:1112
    <anonymous>() at http://theblacksun.s3.amazonaws.com/props/beatMatcher/beatMatcher.js:989
    <global>() at -1

**Comment by hifi-gustavo** [comment]
Build console output for commit: bb7261bd32e121e7938cdc82176ff75a09fb0c2f available at: https://gist.github.com/0ff34e6f9edfb9fcb81800480f0b673e<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9732-bb7261b.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9732-bb7261b.dmg

**Comment by PhilipRosedale** [comment]
Argh...  yeah the PR has relative references to your JS outside the PR.  Can you update the PR to have the createBeatMatcher script reference the beatMatcher script in the PR?   I just want to be testing the right thing.

**Comment by TopRamenGod** [comment]
Fixed!

**Comment by hifi-gustavo** [comment]
Build console output for commit: e86ca3630bdad794fd33f50554257b4861f04883 available at: https://gist.github.com/0ff34e6f9edfb9fcb81800480f0b673e<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9732-e86ca36.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9732-e86ca36.dmg

**Comment by cozza13** [comment]
Hi @TopRamenGod, this is looking good. I have one issue. After a while of it running I get a double up of the text scoreboard 

capture

**Comment by TopRamenGod** [comment]
hi @cozza13 

I think I've figured out what the issue is:
What happens in the current version is that getNewScoreboarDisplay() creates and adds a number of ScoreboardLine entities and objects, and since this is happening in the preload(), andsince this script runs on everyone's client, it will then spawn that number for every person in the domain. This didn't show up until I tested with a bunch of people in the domain, as everything is set to spawn at the original location in front of the orginal who spawned it. 

I'm working on a solution now and should have that pushed soon.

File: scripts/tutorials/entity_scripts/beatMatcher.js [hunk 1]
```diff
@@ -0,0 +1,1237 @@
...
+        this.MISS_COLOR = {color: {red: 255, green: 128, blue: 0}};         // Dark Orange
+
+        // ::: Mouse Click Operation :::
+        this.clickDownOnEntity = function(entityID, mouseEvent){
+            if (Entities.getEntityProperties(theDrumObjThis.entityID).name == "BeatMatcher_Drum") {
+                theDrumObjThis.hitDrum();
+            }
+        };
+
+        rightHandControllerJointIndex = MyAvatar.getJointIndex("_CAMERA_RELATIVE_CONTROLLER_RIGHTHAND");
```
Comment (PhilipRosedale) [inline scripts/tutorials/entity_scripts/beatMatcher.js]: these two variables are assigned but never used.  delete?

Comment (TopRamenGod) [inline scripts/tutorials/entity_scripts/beatMatcher.js]: Yeah, I took care of those and a lot more in my latest pushes below.

**Comment by TopRamenGod** [comment]
@cozza13 The overlapping text should be fixed now.

@PhilipRosedale I ran my code through eslint using the eslintrc.js sitting in the hifi root, which I assume is the prescribed standard. Passes with no errors or warnings, only weak warnings. I think this thing should be ready for submission now.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2ad93d60e231066730d1c7557f665dd39b5ffa32 available at: https://gist.github.com/0ff34e6f9edfb9fcb81800480f0b673e<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9732-2ad93d6.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9732-2ad93d6.dmg

**Comment by cozza13** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: a1b25f1c5d7076a66c9051046d9227770f526ce1 available at: https://gist.github.com/0ff34e6f9edfb9fcb81800480f0b673e<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9732-a1b25f1.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9732-a1b25f1.dmg

**Comment by claudioshg** [comment]
### Desktop PC - Windows 10 Home - CPU 4.0GHz Intel i7 - GTX1070 - RAM 16GB - Tested with Oculus Rift and in Desktop Mode

**Tested:** Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9732-a1b25f1.exe

NOTES:
- Is the intention of the sphere to be hit/clicked on on the off-beat? 
It just seems counter-intuitive not to hit/click the sphere on the beat.
- The spheres detection of my hands were great. Never had any problems with the rhythmic practice toy.
- I had 4 crashes in a span of 1 hour and I couldn't figure out what caused them. It just seemed random.

FAILS:
- No matter how many beats I would match, 3, 56, 105, the final score is always 104.
- Every time the game starts, approximately 1 or 2 seconds in, an image appears and quickly disappears.
The Gif below shows, 2 beats in, an image that says, "Beat Matcher 5000/2017 High Fidelity" show up.
hifi-snap-by-claudioshg-on-2017-03-19_21-47-45

**Comment by ZappoMan** [comment]
@cozza13 @TopRamenGod - what's the status on this PR? are we going to get it fixed/merged?

**Comment by TopRamenGod** [comment]
Hi everyone, had food poisoning and other stuff happen.

@claudioshg 
- the intention is indeed to hit the sphere on beat
- I can't figure out any of the random crashes, nor do I know why the text entities flash the starting screen. My code doesn't flow like that, so I don't know. It's actually designed to 'refresh' the text entity objects with blank text before it updates with a new screen, so I really don't know why it does this.
- _"3, 56, 105, the final score is always 104."_ Absolutely no idea, there is no hard coded limit, or any limit for that matter, on the number of matched beats.

So, at this point I don't know how to fix what's 'broken' without any further insights.

**Pull request closed without merging**